exchanges/coinbasepro/coinbasepro.go

```diff
@@ -3,751 +3,1135 @@ package coinbasepro
+	if productID == "" {
+		return nil, errProductIDEmpty
+	}
+	path := coinbaseV3 + coinbaseProducts + "/" + productID
+	resp := Product{}
+	return &resp, c.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet,
+		path, nil, nil, true, &resp, nil)
 }
 
-// GetTrades listd the latest trades for a product
-// currencyPair - example "BTC-USD"
-func (c *CoinbasePro) GetTrades(ctx context.Context, currencyPair string) ([]Trade, error) {
-	var trades []Trade
-	path := fmt.Sprintf(
-		"%s/%s/%s", coinbaseproProducts, currencyPair, coinbaseproTrades)
-	return trades, c.SendHTTPRequest(ctx, exchange.RestSpot, path, &trades)
+// GetHistoricRates returns historic rates for a product. Rates are returned in
+// grouped buckets based on requested granularity. Requests that return more than
+// 300 data points are rejected
+func (c *CoinbasePro) GetHistoricRates(ctx context.Context, productID, granularity string, startDate, endDate time.Time) ([]CandleStruct, error) {
+	var resp History
+	if productID == "" {
+		return nil, errProductIDEmpty
+	}
+	allowedGranularities := [8]string{granOneMin, granFiveMin, granFifteenMin,
+		granThirtyMin, granOneHour, granTwoHour, granSixHour, granOneDay}
+	validGran, _ := common.InArray(granularity, allowedGranularities)
+	if !validGran {
+		return nil, fmt.Errorf("invalid granularity %v, allowed granularities are: %+v",
+			granularity, allowedGranularities)
+	}
+	vals := url.Values{}
+	vals.Set("start", strconv.FormatInt(startDate.Unix(), 10))
+	vals.Set("end", strconv.FormatInt(endDate.Unix(), 10))
+	vals.Set("granularity", granularity)
+	path := coinbaseV3 + coinbaseProducts + "/" + productID + "/" + coinbaseCandles
+	err := c.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet,
+		path, vals, nil, true, &resp, nil)
+	return resp.Candles, err
 }
 
-// GetHistoricRates returns historic rates for a product. Rates are returned in
-// grouped buckets based on requested granularity.
-func (c *CoinbasePro) GetHistoricRates(ctx context.Context, currencyPair, start, end string, granularity int64) ([]History, error) {
-	values := url.Values{}
+// GetTicker returns snapshot information about the last trades (ticks) and best bid/ask.
+// Contrary to documentation, this does not tell you the 24h volume
+func (c *CoinbasePro) GetTicker(ctx context.Context, productID string, limit uint16, startDate, endDate time.Time) (*Ticker, error) {
+	if productID == "" {
+		return nil, errProductIDEmpty
+	}
+	path := coinbaseV3 + coinbaseProducts + "/" + productID + "/" + coinbaseTicker
+	vals := url.Values{}
+	vals.Set("limit", strconv.FormatInt(int64(limit), 10))
+	if !startDate.IsZero() && !startDate.Equal(time.Time{}) {
+		vals.Set("start", strconv.FormatInt(startDate.Unix(), 10))
+	}
+	if !endDate.IsZero() && !endDate.Equal(time.Time{}) {
+		vals.Set("end", strconv.FormatInt(endDate.Unix(), 10))
+	}
+	var resp Ticker
+	return &resp, c.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet,
```

Comment:
This is not an authenticated endpoint: https://docs.cloud.coinbase.com/advanced-trade-api/reference/retailbrokerageapi_getmarkettrades
`[ERROR]	 | SYNC | 26/03/2024 15:23:23 | Failed to get REST ticker. Error: CoinbasePro REST or Websocket authentication support is not enabled`

```suggestion
	return &resp, c.SendHTTPRequest(ctx, exchange.RestSpot, path, &resp)
```

Reply:
Please verify other API endpoints and verify that they are not attempting to send requests with authentication when it is not needed

Reply:
No V3 endpoint is public except for Get Server Time, as per https://docs.cloud.coinbase.com/advanced-trade-api/docs/rest-api-overview

Although that was accidentally left authenticated too; I'll push a fix for that along with the other stuff you brought up soon.

Reply:
Just to add onto this. Could you please implement all unauthenticated endpoints from https://docs.cloud.coinbase.com/exchange/reference
This will allow for basic data retrieval without credentials. Endpoints such as 
Instruments
Ticker
Orderbook
Trades
Candles

These will be important to reference in the wrapper functions:
`FetchTradablePairs`
`UpdateTickers`
`UpdateTicker`
`UpdateOrderbook`
`GetRecentTrades`
`GetHistoricTrades`
`GetHistoricCandles`
`GetHistoricCandlesExtended`
`UpdateOrderExecutionLimits`